dragonfly/roompolygon.py

```diff
@@ -0,0 +1,421 @@
+from .glazingparameter import GlazingParameter
+
+import honeybee.boundarycondition.boundary_conditions as bcs
+from honeybee.boundarycondition import _BoundaryCondition, Outdoors, Surface
+from honeybee.typing import tuple_with_length, float_positive
+from honeybee.room import Room
+from honeybee.aperture import Aperture
+
+from ladybug_geometry.geometry3d.face import Face3D
+from ladybug_geometry.geometry3d.polyface import Polyface3D
+
+import math
+
+
+class RoomPolygon(_Base):
+    """A volume defined by an extruded floor plate, representing a single room or space.
+
+    Properties:
+        * name
+        * display_name
+        * geometry
+        * floor_to_ceiling_height
+        * boundary_conditions
+        * glazing_ratios
+        * glazing_parameters
+        * parent
+        * has_parent
+        * floor_segments
+        * segment_count
+        * floor_height
+        * ceiling_height
+        * volume
+        * floor_area
+        * exterior_wall_area
+        * exterior_aperture_area
+    """
+    __slots__ = ('_geometry', '_segment_count', '_floor_to_ceiling_height',
+                 '_boundary_conditions', '_glazing_ratios', '_glazing_parameters',
+                 '_parent')
+
+    def __init__(self, name, geometry, floor_to_ceiling_height, boundary_conditions=None,
+                 glazing_ratios=None, glazing_parameters=None, tolerance=None):
+        """A volume defined by an extruded floor plate, representing a single room.
+
+        Args:
+            name: Room name. Must be < 100 characters.
+            geometry: A single Face3D object representing the floor plate of the Room.
+            floor_to_ceiling_height: A number for the height above the floor where the
+                ceiling begins. Typical values range from 3 to 5 meters.
+            boundary_conditions: A list of boundary conditions that match the number of
+                segments in the input Face3D geometry. These will be used to assign
+                boundary conditions to each of the walls of the Room in the resulting
+                model. If None, all boundary conditions will be Outdoors or Ground
+                depending on whether ceiling of the room is below 0 (the assumed
+                ground plane).
+            glazing_ratios: A list of numbers between 0 and 1 that match the number of
+                segments in the input Face3D geometry. These will be used to generate
+                window geometries for the walls in the resulting model. Note that these
+                values must be 0 for any Ground or Adiabatic walls.
+            glazing_parameters: A list of GlazingParameter objects that dictate how the
+                window geometries will be generated from the glazing_ratios. If None,
+                all window geometries will be generated by scaling the parent Face to
+                produce only one window per Face.
+            tolerance: The maximum difference between z values at which point vertices
```

Comment:
what about a non-horizontal but still planar floor? It is valid as long as this input is empty or it is considered invalid?

Reply:
Speaking realistically, I think it's invalid, which is why I plan to always use this input in the Grasshopper plugin. `None` just means that you have performed the check elsewhere and so you don't need to do it again here. FYI, I am using the same logic here with the tolerance as I do with the `Room` `__init__` method and testing whether the volume is closed:
https://github.com/ladybug-tools/honeybee-core/blob/master/honeybee/room.py#L40

Reply:
This has a little bit of magic into it. I feel the tolerance input and enforce check for planarity can be separate.

I also think it will be helpful to be clear that it needs to be a _horizontal_ planar input.

Reply:
I feel it doesn't have any more magic than the check for solidity in the honeybee `Room` volume but you would be right to point out that there's an additional method to check the closure of the Room volume if someone does not initialize the Room with a tolerance. So I'll do the same thing here and add a `check_horizontal` to the `Room2D` that parallels the `check_solid` property of the Room.

I'll also add some extra clarification in the docstring that the floor plane is horizontal.